### ISSUE TITLE
[varLib.mutator] Overlapping contours on old versions of Safari

### ISSUE THREAD

**Issue by m4rc1e** [fonttools/fonttools#1281]
Google Fonts has just released their first VF font, [MarkaziText](https://fonts.google.com/specimen/Markazi+Text). VarLib's mutator has been used to serve static instances to browsers which do not support variable fonts. Unfortunately, users who still use older versions of Safari are [reporting](https://github.com/BornaIz/markazitext/issues/6#issuecomment-399991205) the classic OS X [winding problem](https://github.com/twardoch/test-fonts/tree/master/varia/160413-EvenOddTT).

@garretrieger says that half of all users of Safari are using an outdated version. We could potentially solve this issue by removing overlaps on the mutated instances.

**Comment by behdad** [comment]
See https://github.com/fonttools/fonttools/issues/730

**Comment by behdad** [comment]
cc @lorp

**Comment by Lorp** [comment]
Note that this overlap issue still affects *current* Safari Version 11.1.1 (13605.2.8).

Thanks for the pointer to the "bit 6 hack", @behdad. Please remember that this causes the font to be rejected by [OT sanitizer code](https://github.com/khaledhosny/ots) built into Chrome and Firefox.

Another hack that works is to add a dummy fvar table in the font (e.g. a single custom axis with min, max & default = 0) and no gvar table. I have not seen any side-effects.

cc @twardoch

**Comment by garretrieger** [comment]
Looks like this is was already recently fixed in OTS. Font with the bit flipped works on the latest Chrome/FF. However fails on FF 59 and below and Chrome 63 and below.

**Comment by khaledhosny** [comment]
> Ugh... @khaledhosny can you fix OTS please?

I already did, last year 353c7f83a21ca520f23f7fba480121d0bca3cac9.

**Comment by anthrotype** [comment]
> We could potentially solve this issue by removing overlaps on the mutated instances.

our [fonttools/skia-pathops](https://github.com/fonttools/skia-pathops) can be used to remove overlaps of the mutated instances. It works on quadratic curves as well, and supports the fonttools pen protocol.
We could add an option to fonttools mutator to use that (provided the it's installed).

**Comment by jenskutilek** [comment]
> our fonttools/skia-pathops can be used to remove overlaps of the mutated instances. It works on quadratic curves as well, and supports the fonttools pen protocol.
We could add an option to fonttools mutator to use that (provided pathops is installed).

Removing overlaps might get ugly if the font has TT instructions ...

**Comment by anthrotype** [comment]
should we add an options to fontTools.varLib.mutator to set that flag?

**Comment by anthrotype** [comment]
I meant the OVERLAP_SIMPLE flag for glyf contour points that @jenskutilek added with https://github.com/fonttools/fonttools/pull/1316

**Comment by chrissimpkins** [comment]
https://github.com/fonttools/fonttools/pull/1518 adds support for setting Bit 6 (`OVERLAP_SIMPLE`) of the first flag byte of all glyphs with contour definitions to 1 with `varLib.mutator.instanatiateVariableFont`.

**Comment by Lorp** [comment]
I was going to advise Microsoft to update the `glyf` table spec, explicitly allowing bit 6, but I’m pleased to see that it’s official already :) Thx @PeterCon @robmck-ms
![screenshot 2019-02-24 at 15 01 13](https://user-images.githubusercontent.com/587440/53299627-66329280-3845-11e9-9e89-4f4ec3c14f51.png)

**Comment by khaledhosny** [comment]
I believe this is fixed now.

**Comment by Lorp** [comment]
Note that `woff2_compress` currently sets OVERLAP_SIMPLE to zero. This is undesirable! See [woff2 issue #20](https://github.com/google/woff2/issues/123).

**Comment by anthrotype** [comment]
Yeah, also the fontTools WOFF2 encoder in fontTools.ttLib.woff2 doesn't keep these OVERLAP glyf flags.
However, I believe this is a limitation of the current WOFF2 specification that only encodes the oncurve-ness of a glyph point and throws away any extra flags.
See "5.2. Decoding of variable-length X and Y coordinates" chapter in https://www.w3.org/TR/WOFF2/#triplet_decoding

I guess we could have an option that sets those flags on all glyphs when compressing WOFF2 fonts.

**Comment by anthrotype** [comment]
> I guess we could have an option that sets those flags on all glyphs when compressing WOFF2 fonts.

but we can't, because the format doesn't allow it. The only workaround if one wants to preserve those flags is to _not_ transform the glyf table when compressing the font with WOFF2.

**Comment by anthrotype** [comment]
another way to fix this is have the fonttools instancer run skia-pathops to remove the overlaps so that the OVERALP flags are no longer necessary.

**Comment by anthrotype** [comment]
... however, as @jenskutilek already noted above, removing the overlaps will also invalidate any TrueType hinting that may be present in the instance font.

**Comment by robmck-ms** [comment]
We TT Hint all of our fonts, and really enjoy that those hints still work when a VF is instanced, so it'd be great to keep the hints somehow.

**Comment by Lorp** [comment]
I’m reverting to my earlier solution of an empty `fvar` table to trick Apple to switch renderer :)

**Comment by anthrotype** [comment]
> an empty fvar table to trick Apple to switch renderer

ha! that's a good trick :)
Probably better than having the glyf table un-transformed in woff2 (with the penalty of bigger file size). An empty fvar is only a few bytes.

**Comment by chrissimpkins** [comment]
Btw, this issue was more broad than old Safari versions as the title of this thread suggests when we looked at it in https://github.com/fonttools/fonttools/pull/1518 with static *.ttf fonts that were instantiated from VF's.  It also affected (at least) Chrome on macOS at that time. I don't recall whether we looked into other browsers.  I didn't document others in the PR so I am guessing not.

**Comment by Lorp** [comment]
@chrissimpkins, yes indeed, I should not have specified Safari. It is in fact all native rendering on macOS, so you see the problem in TextEdit and even in Finder font quick view if your A exhibits the problem.

Unfortunately Apple has blocked my `fvar` trick! Safari is rejecting fonts with a present but empty `fvar` table :( This seems to be new (because I tried the trick successfully many months ago), so I guess a recent Safari update can be blamed. I hope other browser vendors do not follow Apple!

**Comment by robmck-ms** [comment]
I worry a bit about the empty fvar trick because that might cause other effects which may be unintuitive or not obvious. e.g. would Windows Chrome switch to render those fonts with freetype (possibly invoking ttfautohint, which I've found produces poor results on fonts with overlaps).

**Comment by Lorp** [comment]
I can confirm that the following tiny `fvar` table (36 bytes uncompressed) survives WOFF2 compression AND passes Safari validation on macOS 10.14, thus successfully nudging macOS to handle overlaps correctly in WOFF2’d TrueTypes:

```
  <fvar>

    <Axis>
      <AxisTag>wwww</AxisTag>
      <Flags>0x0</Flags>
      <MinValue>0.0</MinValue>
      <DefaultValue>0.0</DefaultValue>
      <MaxValue>0.0</MaxValue>
      <AxisNameID>0</AxisNameID>
    </Axis>

  </fvar>
```

Using axis tag `wwww` (= 0x77777777) to optimize compression :) The AxisNameID of 0 is perhaps risky, so keeping to spec and using 256 along with a `name` table update would be sensible.

Good point, Rob, about a specific downside to this platform-specific hack. In our environment we control how to serve the fonts, so we can deliver unmodified static TTF/WOFF2s for Windows Chrome and minimal-fvar versions for Mac.

**Comment by behdad** [comment]
NameID 0xFFFF please.

**Comment by anthrotype** [comment]
As of #1639, you can now compress (and decompress) WOFF2 without glyf transformation. Up to you to decide whether to use the trick suggested by @Lorp above (adding a dummy fvar table) or just compress as un-transformed WOFF2.
I think we can close this now.

**Comment by bavan123** [comment]
also you can try [Cool Fancy Fonts](https://coolfancyfonts.com) Tool To write messages in (cool, Fancy, stylish) unusual fonts For any uses
